### ISSUE TITLE
Building a RESTful API for enabling use of image sequencer on a remote server

### ISSUE THREAD

**Issue by tech4GT** [publiclab/image-sequencer#198]
Some Image processing steps can be resource intensive and it might be difficult for users with low powered devices to run them locally one the device, for this purpose we can setup an Express server with an api that can take an image in request and return the step applied image in response.
For this purpose we can use https://www.npmjs.com/package/express for building the server and imgur for hosting the images. The image at all steps will be uploaded to imgur. https://imgur.com
We can start very basic with support for a single step and later add functionalities add multiple steps and replace steps
The demo will have a choice if user want to run on local machine or remote server for each step, we can also limit this in the basis of user's computer specification to prevent overloading the server

**Comment by jywarren** [comment]
[GitMate.io](https://gitmate.io) thinks the contributors most likely able to help are @ccpandhare, and @jywarren.

Possibly related issues are https://github.com/publiclab/image-sequencer/issues/173 (Building out a contributor community for Image Sequencer), https://github.com/publiclab/image-sequencer/issues/23 (Planning Issue for the Image Sequencer GSoC Project), and https://github.com/publiclab/image-sequencer/issues/99 (Make the API Documentation using Swagger).

**Comment by tech4GT** [comment]
@jywarren please have a look, this is pretty much what i had in mind, for summer maybe we can just start with support for single step in the server, it would be relatively simpler project and would take 1-2 weeks to implement?

**Comment by ccpandhare** [comment]
@tech4GT a very good idea indeed! But we need to ensure that only those
steps which will take more time are routed to the server. We don't want the
user to be waiting! And how will we decide which is faster? Local or server
(based on internet speed, user configuration, etc.?)

I think this will take more than a week or two... Including testing,
bugfixes and improvements. Usually this takes more time than the
development itself but it is very important! Maybe you can come up with a
timeline?

**Comment by tech4GT** [comment]
@ccpandhare actually i was thinking we can have the user decide if he wants to use local device or server but i get it, this can take a little more time now working on the timeline, thanks👍

**Comment by tech4GT** [comment]
@jywarren @ccpandhare would this be a priority or #190  for summer coz @ccpandhare you are right i cannot do both in the given timeframe, maybe we can work on one after the summer, so should i keep this one on #190 in my proposal? thanks

**Comment by jywarren** [comment]
I think it'd be really useful for example to implement a completely separate service-based MapKnitter back-end, which could receive full-size images and distort them remotely for people. https://github.com/publiclab/mapknitter/

But I think #190 and anything that can help onboard new people to grow this team is a higher priority -- i've written a bit about my reasoning for this here -- hope that makes sense!

https://publiclab.org/notes/tech4gt/02-19-2018/gsoc-proposal-image-sequencer-v2-processing-on-steroids#c18878

**Comment by jywarren** [comment]
Getting back to this one, esp after all the great improvements this last summer. 

This could involve setting up a new repository in which we have a containerized/dockerized node version of image-sequencer which can accept commands via REST API calls.

The string syntax makes this MUCH easier, by allowing us to pass an image or url plus an instruction set. An image would be output.

Another thing we could offer is a callback url maybe!

So if we imagined what a POST request payload would look like, it might include:

```
// Request:

{
  img_src: "...",
  callback: "https://...",
  commandString: "blur,brightness..."
}


// Response:

{
  progress_url: "https://...",
  status: "success"
}
```

**Comment by tech4GT** [comment]
@jywarren This looks great! I didn't quite get the purpose for callback though. Also I was thinking of starting this one in my winter break.

**Comment by jywarren** [comment]
The callback is not necessary for most cases. But it could function like a
push request back to an originating server which could replace polling for
progress. Just thinking ahead to useful architecture.

**Comment by tech4GT** [comment]
Hmm, I am thinking rather than doing it this way maybe we can use web sockets for this kind of functionality, what do you think?

**Comment by jywarren** [comment]
I think websockets could be cool, but a basic API may be easier to
implement and integrate -- the client may not be submitting from javascript
in any case. So perhaps we could work on a RESTful approach first, but then
consider if a websockets interface is a good follow-up! What do you think?

**Comment by jywarren** [comment]
This could relate to https://github.com/publiclab/pi-builder/pull/42 where we want to run Image Sequencer on the Raspberry Pi!

**Comment by jywarren** [comment]
@icarito i'm a little curious about this - do you think we could containerize this library and run a "job" processing a batch of image sequencer steps on a batch of images, in a CI process? Such that we could "submit" a job with URLs of a set of images, and get out the processed output as an artifact?

**Comment by Mridul97** [comment]
This issue seems interesting to me. Is anyone working on this issue? Can I work on this issue?

**Comment by VibhorCodecianGupta** [comment]
@Mridul97 I was working on this. Would you like to collaborate?

**Comment by Mridul97** [comment]
Yes, sure!

**Comment by VibhorCodecianGupta** [comment]
@Mridul97 amazing! Give me a couple of hours, I'll update you on whatever I have managed to get done and commit it as well. Not at home right now. Is that fine?

**Comment by Mridul97** [comment]
Yes. That's fine!

**Comment by Mridul97** [comment]
@VibhorCodecianGupta I have created a branch building_a_RESTful_API and invited you to collaborate, you can push your commits there and I will push mine as well!

**Comment by VibhorCodecianGupta** [comment]
I just accepted, sweet! Will do by tonight 👍

**Comment by jywarren** [comment]
this is so cool! One thing we were thinking of here is to build a
Dockerfile around this library. I wonder if the RESTFUL API code would best
be kept quite separate from the main codebase - like a wrapper around it --
please keep this in mind as you're working on it! Thanks!

**Comment by Mridul97** [comment]
Yes, I agree it will be better if we keep the RESTFUL API code separate from main codebase!

**Comment by VibhorCodecianGupta** [comment]
@jywarren agreed.

@Mridul97 shall we decide the architecture then?
I think initialising a separate repository for this project would be appropriate, as @jywarren suggested a few comments above. That can include both the express app and the dockerized container for image sequencer. Thoughts?

**Comment by Mridul97** [comment]
It sounds good! So let's start by first creating it for single step and then improve it from there!

**Comment by VibhorCodecianGupta** [comment]
Yes, sounds good and viable!

**Comment by jywarren** [comment]
This is awesome. If you decide on a project name -- maybe
'image-sequencer-app'? Or something else? -- I can create a `publiclab`
repository we can use as the root. Sound good?

**Comment by VibhorCodecianGupta** [comment]
Sounds good!

**Comment by Mridul97** [comment]
Yes, Sounds good!

**Comment by jywarren** [comment]
Great! Also let's keep in mind the ability to run serverless... Would this be possible in the scenario we've planned out?

https://i-programmer.info/news/141-cloud-computing/12405-gitlab-goes-serverless.html

**Comment by jywarren** [comment]
https://github.com/publiclab/image-sequencer-app

**Comment by jywarren** [comment]
And just a note that the upload to imgur needn't be implemented in the first version. What's the most standard way for a serverless API to return an image?

**Comment by jywarren** [comment]
https://serverless.com/

**Comment by tech4GT** [comment]
This is awesome!! This can also serve as a very good example use case to image sequencer in node.js.

**Comment by tech4GT** [comment]
@Mridul97 @VibhorCodecianGupta I think you guys can start of with a simple express based server where you get a stringified sequence in the request and you can expand it and run it one the backend. We can then send the final image either as multipart data or the imgur url

**Comment by HarshKhandeparkar** [comment]
That's good. What I actually wanted to ask is how will the user get the preview for intermediate steps even from the API? This will defeat the purpose of image sequencer's non-conventional workflow in which a new image is created every step.

**Comment by HarshKhandeparkar** [comment]
I'm just curious and because I love the idea of RESTful APIs.

**Comment by jywarren** [comment]
I think that for the initial implementation we can return just the final
image. But later, we could return a list of URLs for the image at each
step.

Will Express be compatible with the emerging "serverless" format? Just
checking, 'cause that'd be cool!

**Comment by jywarren** [comment]
Oohhhhhh!!! 👍🏽

**Comment by jywarren** [comment]
Just collecting resources here... We might be able to use Google cloud to store images, or Amazon s3. I think we may have a preference for google cloud however as we can potentially get a discount. 

https://dzone.com/articles/serverless-image-processing-patterns-with-aws

**Comment by jywarren** [comment]
And this mentions a template docker image called lambci/lambda for Amazon: https://read.acloud.guru/serverless-image-optimization-and-delivery-510b6c311fe5

**Comment by jywarren** [comment]
https://github.com/awslabs/serverless-application-model/tree/master/examples/apps/image-processing-service

**Comment by HarshKhandeparkar** [comment]
@jywarren how about [firebase](https://firebase.google.com)

**Comment by HarshKhandeparkar** [comment]
Hmm.. then I guess GCP is the best option

**Comment by HarshKhandeparkar** [comment]
Firebase is probably better for like a blog or something

**Comment by jywarren** [comment]
Just pasting in more documentation here - https://www.thepolyglotdeveloper.com/2017/12/convert-nodejs-restful-api-serverless-openwhisk/

**Comment by jywarren** [comment]
Ah here, maybe? https://cloud.google.com/functions/use-cases/real-time-data-processing
